File: beginner/python/anagrams.py
def is_anagram(word1, word2):
  if len(word1) != len(word2):
    return False
  
  word_one_sorted = ''.join(sorted(word1))
  word_two_sorted = ''.join(sorted(word2))

  return word_one_sorted == word_two_sorted

def is_anagram_w_char_map(word1, word2):
  if len(word1) != len(word2):
    return False
  
  def create_char_map(text):
    char_map = {}

    for char in text:
      if char not in char_map:
        char_map[char] = 1
      else:
        char_map[char] += 1
    
    return char_map

  word1_char_map = create_char_map(word1)
  word2_char_map = create_char_map(word2)

  for char in word1_char_map:
    if char in word2_char_map:
      if word1_char_map[char] != word2_char_map[char]:
        return False
    else:
      return False
  
  return True



print(is_anagram_w_char_map('seen', 'needless'))
print(is_anagram_w_char_map('silent', 'listen'))
print(is_anagram_w_char_map('two', 'far'))

File: beginner/typescript/anagrams.ts
function isAnagram(word1: string, word2: string): boolean {
  if (word1.length !== word2.length) {
    return false;
  }
  const sorted1 = word1.toLowerCase().split('').sort().join('');
  const sorted2 = word2.toLowerCase().split('').sort().join('');
  return sorted1 === sorted2;
}

interface CharMap {
  [key: string]: number;
}

function isAnagramCharMap(word1: string, word2: string): boolean {
  if (word1.length !== word2.length) {
    return false;
  }

  function createCharMap(text: string): CharMap {
    const charMap: CharMap = {};

    for (let char of text) {
      if (!charMap[char]) {
        charMap[char] = 1;
      } else {
        charMap[char]++;
      }
    }

    return charMap;
  }

  const word1CharMap = createCharMap(word1);
  const word2CharMap = createCharMap(word2);

  for (let char in word1CharMap) {
    const word1CharCount = word1CharMap[char];
    const word2CharCount = word2CharMap[char];
    if (!word2CharCount || word2CharCount !== word1CharCount) {
      return false;
    }
  }

  return true;
}

console.log(isAnagramCharMap('silent', 'listen'));
console.log(isAnagramCharMap('seen', 'needless'));
console.log(isAnagramCharMap('two', 'far'));
